paparazzi-preview-processor/src/main/java/app/cash/paparazzi/preview/processor/PaparazziPoet.kt

```diff
@@ -0,0 +1,144 @@
+package app.cash.paparazzi.preview.processor
+
+import com.google.devtools.ksp.getVisibility
+import com.google.devtools.ksp.symbol.KSFunctionDeclaration
+import com.google.devtools.ksp.symbol.KSValueParameter
+import com.google.devtools.ksp.symbol.Visibility
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.buildCodeBlock
+
+internal object PaparazziPoet {
+  fun buildFiles(
+    functions: Sequence<KSFunctionDeclaration>,
+    isTest: Boolean,
+    env: EnvironmentOptions
+  ) =
+    if (isTest) {
+      emptyList()
+    } else {
+      listOf(
+        buildAnnotationsFile(
+          fileName = "PaparazziPreviews",
+          propertyName = "paparazziPreviews",
+          functions = functions,
+          env = env
+        )
+      )
+    }
+
+  @Suppress("SameParameterValue")
+  private fun buildAnnotationsFile(
+    fileName: String,
+    propertyName: String,
+    functions: Sequence<KSFunctionDeclaration>,
+    env: EnvironmentOptions
+  ) =
+    FileSpec.scriptBuilder(fileName, env.namespace)
+      .addCode(
+        buildCodeBlock {
+          addStatement("internal val %L = listOf<%L.PaparazziPreviewData>(", propertyName, PACKAGE_NAME)
+          indent()
+
+          if (functions.count() == 0) {
+            addEmpty()
+          } else {
+            functions.process { func, previewParam ->
+              val visibilityCheck = checkVisibility(func)
+              val snapshotName = func.snapshotName(env)
+
+              when {
+                visibilityCheck.isPrivate -> addError(
+                  function = func,
+                  snapshotName = snapshotName,
+                  buildErrorMessage = {
+                    "$it is private. Make it internal or public to generate a snapshot."
+                  }
+                )
+                previewParam != null -> addError(
+                  function = func,
+                  snapshotName = snapshotName,
+                  buildErrorMessage = {
+                    "$it preview uses PreviewParameters which aren't currently supported."
+                  }
+                )
```

Comment:
propagating error and empty states seem confusing and goes against the "fail fast" principle.  Let's make these compile-time errors and add lint checks to educate consumers?